<a href="https://colab.research.google.com/github/liperdio/liperdio.github.io/blob/master/MLP_for_CIFAR10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
from keras.datasets import cifar10

In [0]:
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

# numpy package
import numpy as np
from keras.models import Sequential
from keras.layers import Dense, Activation,Flatten, Dropout
from keras.datasets import mnist
from keras.regularizers import l2
from keras.utils import to_categorical

In [0]:
#Load Cifar10 dataset
(x_train, y_train_), (x_test, y_test_) = cifar10.load_data()

In [0]:
#Resize
x_train = np.reshape(x_train, [50000, 32*32*3])
x_test = np.reshape(x_test, [10000, 32*32*3])

In [0]:
#Compute the number of labels
num_labels = len(np.unique(y_train_))

In [0]:
#Normalize
x_train = x_train.astype('float32') / 255
x_test = x_test.astype('float32') / 255

In [0]:
#Convert to one-hot vector
from keras.utils import to_categorical
y_train = to_categorical(y_train_)
y_test = to_categorical(y_test_)

In [0]:
#Network Parameters
batch_size = 128
hidden_units = 256

kernel_regularizer = l2(0.0001)

In [0]:
#A Dense-dense-dense-dense-activation model 
from keras.models import Sequential
model = Sequential()

In [71]:

model.add(Dense(hidden_units,
                kernel_regularizer=kernel_regularizer,
                input_dim=3072))
model.add(Activation('relu'))
model.add(Dense(hidden_units,
                kernel_regularizer=kernel_regularizer))
model.add(Activation('relu'))
model.add(Dense(hidden_units,
                kernel_regularizer=kernel_regularizer))
model.add(Activation('relu'))
model.add(Dense(num_labels))
# this is the output for one-hot vector
model.add(Activation('softmax'))
model.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_21 (Dense)             (None, 256)               786688    
_________________________________________________________________
activation_21 (Activation)   (None, 256)               0         
_________________________________________________________________
dense_22 (Dense)             (None, 256)               65792     
_________________________________________________________________
activation_22 (Activation)   (None, 256)               0         
_________________________________________________________________
dense_23 (Dense)             (None, 256)               65792     
_________________________________________________________________
activation_23 (Activation)   (None, 256)               0         
_________________________________________________________________
dense_24 (Dense)             (None, 10)               

In [0]:
#use of adam optimizer
#loss function for one-hot vector

model.compile(optimizer='adam',
             loss='categorical_crossentropy',
             metrics=['accuracy'])

In [73]:
# train the network
history = model.fit(x_train, y_train, batch_size=128, epochs=20)

Epoch 1/20
50000/50000 [==============================] - 10s 197us/step - loss: 1.9264 - acc: 0.3263
Epoch 2/20
50000/50000 [==============================] - 9s 188us/step - loss: 1.7083 - acc: 0.4052
Epoch 3/20
50000/50000 [==============================] - 9s 187us/step - loss: 1.6130 - acc: 0.4375
Epoch 4/20
50000/50000 [==============================] - 9s 186us/step - loss: 1.5535 - acc: 0.4561
Epoch 5/20
50000/50000 [==============================] - 9s 188us/step - loss: 1.5178 - acc: 0.4726
Epoch 6/20
50000/50000 [==============================] - 9s 185us/step - loss: 1.4811 - acc: 0.4873
Epoch 7/20
50000/50000 [==============================] - 9s 186us/step - loss: 1.4553 - acc: 0.4922
Epoch 8/20
50000/50000 [==============================] - 9s 185us/step - loss: 1.4330 - acc: 0.5054
Epoch 9/20
50000/50000 [==============================] - 9s 187us/step - loss: 1.4104 - acc: 0.5141
Epoch 10/20
50000/50000 [==============================] - 9s 189us/step - loss: 1.3852 - 

In [74]:
# validate the model on the test dataset to determine generalization
loss, acc = model.evaluate(x_test, y_test, batch_size=128)
print("\nTest accuracy: %.1f%%" % (100.0 * acc))


10000/10000 [==============================] - 1s 64us/step

Test accuracy: 51.7%


The dataset used for the experiment is the CIFAR10 dataset which consists of 50,000 32x32 pixel color training images, labeled over 10 categories, and 10,000 test images.

MLP has the characteristic of fully connected layers, where each perceptron is connected with every other perceptron. After evaluating MLP model against the data it has never seen before (test set), it achieved an accuracy of 51.7%. The low accuracy shows one of the disadvantages of MLP for images which is that it disregards spatial information as it takes flattened vector as inputs.

On the other hand, CNN allows panning of filters which enables parameter and weight sharing, where the filter could find a specific pattern anywhere in the image. CNN layers are sparsely connected rather than fully connected. After evaluation against the test set, achieving 73.8% accuracy shows that CNN has a sheer conservation of dimensional data which is very crucial in recognizing images such as CIFAR10 dataset.

MLP could be used on the CIFAR10 dataset, but the CNN is more suited for the experiment.